### ISSUE TITLE
False Problem UnnecessaryVarAnnotation

### ISSUE THREAD

**Issue by mj4444ru** [vimeo/psalm#9721]
https://psalm.dev/r/f34213f312

Because of the line `$viewName ??= 'password-view';` does psalm think the condition `$viewName === 'password-logs'` is always false?

**Comment by ygottschalk** [comment]
Nope, psalm would emitt a `TypeDoesNotContainType` error (see: https://psalm.dev/r/99356d9ace).

You can also verify this with your given code sample. Open it on psalm.dev and hover over `$vars` with your mouse in the last line of the code. Notice the `logRows?` key in the array shape.

Psalm just does not like the comment / type hints / variable declarations in line 32-34. Workaround: use the mighty semicolon after the comment (see: https://psalm.dev/r/de95ef880c). Don't ask me why, it just sometimes works...

**Comment by mj4444ru** [comment]
> Workaround: use the mighty semicolon after the comment

Thanks for the semicolon tip.

**Comment by mj4444ru** [comment]
I think it would be good to add a description of this problem and its solution to page https://psalm.dev/docs/running_psalm/issues/UnnecessaryVarAnnotation/

**Comment by mj4444ru** [comment]
```php
/**
 * @var PasswordService $passwordService
 * @var mixed $passwordKey
 * @var array{id: int} $password
 */;
```

PhpStorm reports an error with an extra semicolon. And this is correct, since this error can really be, but the fact that the semicolon comes after the comment is not taken into account. When we remove one problem, we create another.
The behavior of the code analyzer should definitely be fixed here.

**Comment by ygottschalk** [comment]
> PhpStorm reports an error with an extra semicolon.

Adding a `;` is a perfectly valid noop, not an error. (See [php-src](https://github.com/php/php-src/blob/master/Zend/zend_language_parser.y#L525))

> I don't think removing the semicolon [...] should have this effect.

Agree. Guessing that the Docblock is consumed differently in the AssignmentAnalyzer vs the "NoopAnalyzer".

> I think it would be good to add a description of this problem and its solution to page

I disagree. This kind of "declaring" a variable seems to be a workaround for something. Hard to tell where the variables would come from in this specific example, but I am guessing they are declared in `_base-password-view.php`. If so, I would suggest to return those variables from that script and assign them in the given one like this: https://psalm.dev/r/5c2045fd34
Don't know if psalm can type those variables on their own based on the return types of an included script (in example the first version). If not, you can type those on their declaration/assignment (version 2).

**Comment by ygottschalk** [comment]
Also, this might help: [Configuration - skipChecksOnUnresolvableIncludes](https://psalm.dev/docs/running_psalm/configuration/#skipchecksonunresolvableincludes)
But I have never used that so I do not know how this setting works exactly...
